File: ex155.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.r = []
        

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.r.append(x)
        

    def pop(self):
        """
        :rtype: void
        """
        self.r.pop(-1)
        

    def top(self):
        """
        :rtype: int
        """
        return self.r[-1]
        

    def getMin(self):
        """
        :rtype: int
        """
        return min(self.r)


minStack = MinStack()
minStack.push(-2)
minStack.push(0)
minStack.push(-3)
minStack.getMin()
minStack.pop()
minStack.top()
minStack.getMin()

File: code/ex155.cpp
// Hash Map

class MinStack {
public:
    stack<int> s;
    int MIN = INT_MAX, cnt = 0;
    unordered_map<int, int> map;    // Save the minimum number while the stack has i items
    
    /** initialize your data structure here. */
    MinStack() {
        map[0] = INT_MAX;
    }
    
    void push(int x) {
        s.push(x);
        MIN = min(MIN, x);
        map[++cnt] = MIN;
    }
    
    void pop() {
        s.pop();
        map.erase(cnt--);
        MIN = map[cnt];
    }
    
    int top() {
        return s.top();
    }
    
    int getMin() {
        return map[cnt];
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */